# Identification des utilisateurs de Copilote

**MOD 7.2 (Introduction à la sciences des données)**, BE séances 1, 2, 3

**Enseignants :** Julien Velcin (CM, BE), Erwan Versmée (BE)

Le projet qu'on vous demande de réaliser concerne le traitement de **traces d'utilisation** du logiciel Copilote. Ce logiciel édité par la société Infologic, concerne le secteur de l'agro-alimentaire et est utilisé par de nombreux acteurs industriels.

L'étude de ces traces est très important pour l'entreprise. Elle permet entre autre d'améliorer son offre auprès de ses clients. Le projet vous propose de travailler sur des **modèles d'analyse automatique** de ces traces.

## Consignes

### Délimitation du travail

Pour ce projet, vous travaillerez par groupe de 3 élèves. Des groupes plus petits peuvent être autorisés même s'ils ne sont pas conseillés.

Les 3 séances doivent vous permettre d'avancer dans la résolution du problème qui est présenté sur la [page Kaggle de la compétition](https://www.kaggle.com/competitions/qui-utilise-mon-appli-challenge/). Elles seront grosso modo organisées de la manière suivante :

1. Découverte des données, construction des variables, mise en forme tabulaire, nettoyage, premières explorations et visualisations
2. Explorations plus avancées des données, développement d'un premier modèle de classification automatique
3. Optimisation de votre solution de classification automatique, participation au challenge

A l'issue de ces trois séances, vous aurez un peu plus d'une semaine supplémentaire pour compléter ce travail.

Date limite de rendu : **2/11/25 au soir**

### Ce qui est attendu

Le rendu consiste en 3 éléments :

- le *code Python* de votre solution, code qui doit être ré-exécutable dans un autre environnement (à priori, python en version 3.13). N'oubliez pas de fournir également les bibliothèques nécessaires par ex. via un fichier *requirements.txt*
- un *rapport* au format PDF de 5 à 8 pages qui doit présenter succinctement votre travail, les choix que vous avez faits et leurs motivations, les résultats obtenus.
- le *dépôt* de vos prédictions sur l'ensemble de test du challenge.

Le **rapport** doit être déposé sur la page moodle de l'enseignement. Il doit être au format PDF et faire entre 5 et 8 pages. Il présente le problème que vous attaquez et comment vous proposez d'y répondre. Il faut en particulier présenter les différentes caractéristiques choisies, si nécessaire avec une petite explication, et les approches de classification mises en oeuvre. Il faut bien sûr présenter les résultats obtenus, en précisant bien les métriques employées, identifier si possible les variables les plus influentes, mener une petite discussion, avant de conclure.

Le **code** doit être suffisamment bien structuré et commenté pour qu'un observateur en comprenne rapidement le fonctionnement, y compris à la lumière du rapport. N'hésitez pas à le rendre disponible sur une plateforme de type git (avec le fichier README adapté) et de fournir l'url directement dans le rapport. Sinon, vous pouvez aussi fournir le code dans une archive (ZIP, TAR...) et le déposer en même temps que votre rapport, mais *sans* y mettre les fichiers de données.

Il est possible de rendre un notebook *à condition* que celui-ci soit succinct et très lisible (ce qui ne veut pas dire que le code intégral soit, lui, succinct).

Le **dépôt** doit être réalisé sur le page du leaderboard de la compétition :

- Groupe TD - 1 (8h-10h), Julien Velcin : https://www.kaggle.com/t/077eb1959d124b11971668f381767f06
- Groupe TD - 2 (8h-10h), Erwan Versmée :  https://www.kaggle.com/t/2374f58cda734c69a28a8ba3ef6692e8
- Groupe TD - 3 (10h-12h),  Julien Velcin :  https://www.kaggle.com/t/ce3349734cc44028bad2875d462e7085
- Groupe TD - 4 (10h-12h), Erwan Versmée :  https://www.kaggle.com/t/6435d33bed324430a9749337b04c2168

### Evaluation

L'évaluation de ce travail sera réalisée selon trois critères :

- mise en place d'une chaîne de traitement complète d'analyse
- capacité à expliquer et motiver les différents traitement mis en oeuvre
- production d'une solution efficace dans un contexte concurrentiel

En plus d'une note /20, vous serez également évalué lors de la 3ème séance de BE selon la compétence C2C2 **Résoudre et arbitrer** :

- *A (remarquable si)* : identifie les méthodes les plus pertinentes pour répondre au problème imposé, exploite la méthode choisie pour répondre au problème, choisit des critères pertinents pour évaluer l'efficacité.
- *C (acquis si)* : identifie les méthodes les plus pertinentes pour répondre au problème imposé, exploite la méthode choisie pour répondre au problème.
- *F (à travailler si)* : n'identifie pas les méthodes les plus pertinentes pour répondre au problème imposé, et/ou propose une exploitation de la méthode choisie qui ne répond pas suffisamment au problème.

## Résolution du problème

Nous vous proposons de suivre les étapes suivantes dans le traitement de cette problématique :

- Prendre connaissance du problème et des données à votre disposition
- Chargement des données brutes en mémoire
- Première analyse de ces données
- Construction de caractéristiques adaptées
- Statistiques simples sur les variables construites
- Développement d'une solution de classification automatique
- Evaluation de votre solution
- Discussion au sujet de vos résultats


## Prendre connaissance du problème et des données à votre disposition

Pour commencer, allez lire la description plus détaillée du problème qui vous est donnée sur la page du challenge mis en place cette année (cf. lien ci-dessus).

Les données sont présentées plus en détail et téléchargeables via l'onglet "data".

L'objectif consiste à comprendre à quoi correspondent chacune des lignes du tableau des données et commencer à vous poser des questions sur la manière d'utiliser ces données pour répondre au challenge. Pour le moment, vous pouvez utiliser un **simple éditeur** (comme Notepad++, Sublime Text...) pour aller voir le contenu des fichiers, les importer dans Excel (mais attention car le volume peut rendre les manipulatins difficiles), ou passer déjà par Python si vous êtes suffisamment à l'aise.

## Chargement des données brutes en mémoire

La toute première étape dans le traitement des données avec Python est de les charger en mémoire pour pouvoir mieux les examiner. Plusieurs solutions s'offrent alors à vous, parmi lesquelles :

- utiliser la librairie Pandas et les méthodes qui permettent de charger directement les données dans des DataFrame comme *readcsv*,
- ouvrir le fichier texte en mode lecture (*open*) et parcourir les lignes grâce à la commande *readlines()*

Cette dernière approche semble plus robuste au regard de la régularité du fichier d'entrée. Identifiez le délimiteur qui permet de séparer les différentes actions de l'utilisateur et utilisez-le pour construire la liste des actions réalisées.

In [27]:
import pandas as pd
import os
import csv

def read_ds(ds_name: str):
    """Robust CSV loader for files where rows have variable field counts.
    Uses Python's csv.reader to parse lines, pads rows to the maximum column count with empty strings,
    and returns a pandas DataFrame with missing values replaced by empty strings.
    Accepts either 'train' or 'train.csv' as input.
    """
    filename = ds_name if ds_name.endswith('.csv') else ds_name + '.csv'
    path = os.path.join('data', filename)
    # Read using csv.reader which avoids pandas C-engine tokenization errors on malformed rows
    with open(path, 'r', encoding='utf-8', errors='replace') as f:
        reader = csv.reader(f)
        rows = [row for row in reader]
    if not rows:
        print(f"No rows read from {path}")
        return pd.DataFrame()
    max_cols = max(len(r) for r in rows)
    padded = [r + [''] * (max_cols - len(r)) for r in rows]
    df = pd.DataFrame(padded)
    df = df.fillna('')
    rename_map = {col: ('util' if col == 0 else 'browser' if col == 1 else f'action_{col-1}') for col in df.columns}
    df.rename(columns=rename_map, inplace=True)
    return df

# Example call (accepts with or without .csv)
read_ds('train.csv')

util         browser                                           action_1  \
0     nuh         Firefox  Création d'un écran(infologic.core.accueil.Acc...   
1     muz   Google Chrome  Création d'un écran(infologic.core.gui.control...   
2     zrx  Microsoft Edge  Affichage d'une dialogue(infologic.core.gui.co...   
3     pou         Firefox  Création d'un écran(infologic.core.gui.control...   
4     ald   Google Chrome  Affichage d'une dialogue(infologic.acti.module...   
...   ...             ...                                                ...   
3274  muz   Google Chrome  Création d'un écran(infologic.core.gui.control...   
3275  cjr   Google Chrome  Affichage d'une dialogue(infologic.acti.module...   
3276  fuz         Firefox  Sélection d’un onglet(infologic.core.gui.contr...   
3277  cjr   Google Chrome  Création d'un écran(infologic.core.gui.control...   
3278  fwf         Firefox  Affichage d'une dialogue(infologic.orga.module...   

                                               action_2  \
0                              Affichage d'une dialogue   
1     Création d'un écran(infologic.core.gui.control...   
2                                 Exécution d'un bouton   
3                                                    t5   
4                                                    t5   
...                                                 ...   
3274  Création d'un écran(infologic.core.gui.control...   
3275                              Exécution d'un bouton   
3276                   Dissimulation d'une arborescence   
3277                                                 t5   
3278                               Saisie dans un champ   

                                               action_3  \
0                                 Exécution d'un bouton   
1                                                    t5   
2                                              Chainage   
3                          Exécution d'un bouton(MAINT)   
4                                 Exécution d'un bouton   
...                                                 ...   
3274                                                 t5   
3275                           Fermeture d'une dialogue   
3276                   Dissimulation d'une arborescence   
3277  Création d'un écran(infologic.core.gui.control...   
3278                              Exécution d'un bouton   

                                               action_4  \
0                              Fermeture d'une dialogue   
1     Sélection d’un onglet(infologic.orga.modules.O...   
2                              Fermeture d'une dialogue   
3                              Affichage d'une dialogue   
4                              Fermeture d'une dialogue   
...                                                 ...   
3274  Sélection d’un onglet(infologic.orga.modules.O...   
3275                                                 t5   
3276                              Exécution d'un bouton   
3277                                                t10   
3278                           Fermeture d'une dialogue   

                                               action_5  \
0                              Affichage d'une dialogue   
1                                                   t10   
2     Affichage d'une dialogue(infologic.acti.module...   
3                              Fermeture d'une dialogue   
4                                                   t10   
...                                                 ...   
3274                               Saisie dans un champ   
3275  Sélection d'un écran(infologic.crm.modules.CRM...   
3276                                           Chainage   
3277                          Création d'un écran(GOUV)   
3278                                                 t5   

                                           action_6  \
0                             Exécution d'un bouton   
1                             Exécution d'un bouton   
2     Clic sur une grille d'historique de r

In [28]:
features_train = read_ds("train")
features_test = read_ds("test")
features_train.shape, features_test.shape

((3279, 14470), (324, 7726))

In [30]:
features_train.head()

util         browser                                           action_1  \
0  nuh         Firefox  Création d'un écran(infologic.core.accueil.Acc...   
1  muz   Google Chrome  Création d'un écran(infologic.core.gui.control...   
2  zrx  Microsoft Edge  Affichage d'une dialogue(infologic.core.gui.co...   
3  pou         Firefox  Création d'un écran(infologic.core.gui.control...   
4  ald   Google Chrome  Affichage d'une dialogue(infologic.acti.module...   

                                            action_2  \
0                           Affichage d'une dialogue   
1  Création d'un écran(infologic.core.gui.control...   
2                              Exécution d'un bouton   
3                                                 t5   
4                                                 t5   

                       action_3  \
0         Exécution d'un bouton   
1                            t5   
2                      Chainage   
3  Exécution d'un bouton(MAINT)   
4         Exécution d'un bouton   

                                            action_4  \
0                           Fermeture d'une dialogue   
1  Sélection d’un onglet(infologic.orga.modules.O...   
2                           Fermeture d'une dialogue   
3                           Affichage d'une dialogue   
4                           Fermeture d'une dialogue   

                                            action_5  \
0                           Affichage d'une dialogue   
1                                                t10   
2  Affichage d'une dialogue(infologic.acti.module...   
3                           Fermeture d'une dialogue   
4                                                t10   

                                        action_6                  action_7  \
0                          Exécution d'un bouton  Fermeture d'une dialogue   
1                          Exécution d'un bouton                       t15   
2  Clic sur une grille d'historique de recherche                 Raccourci   
3                                    Double-clic     Exécution d'un bouton   
4            Entrée en saisie dans un formulaire                       t10   

                                            action_8  ... action_14459  \
0  Création d'un écran(infologic.core.gui.control...  ...                
1                              Sélection d’un onglet  ...                
2                           Fermeture d'une dialogue  ...                
3  Lancement d'une stat(infologic.core.gui.contro...  ...                
4                           Affichage d'une dialogue  ...                

  action_14460 action_14461 action_14462 action_14463 action_14464  \
0                                                                    
1                                                                    
2                                                                    
3                                                                    
4                                                                    

  action_14465 action_14466 action_14467 action_14468  
0                                                      
1                                                      
2                                                      
3                                                      
4                                                      

[5 rows x 14470 columns]

## Première analyse de ces données

Les données contiennent l'utilisateur (e.g. la variable dépendante, uniquement pour l'ensemble d'apprentissage), le navigateur choisi par l'utilisateur, puis toutes les actions effectuées dans la session par l'utilisateur actuel. Un marqueur spécial « tXX » indique un intervalle de temps de 5 secondes.

In [ ]:
features_train.loc[:, ['util','browser', 'action_1', 'action_2', 'action_3']].head()

util         browser                                           action_1
0  nuh         Firefox  Création d'un écran(infologic.core.accueil.Acc...
1  muz   Google Chrome  Création d'un écran(infologic.core.gui.control...
2  zrx  Microsoft Edge  Affichage d'une dialogue(infologic.core.gui.co...
3  pou         Firefox  Création d'un écran(infologic.core.gui.control...
4  ald   Google Chrome  Affichage d'une dialogue(infologic.acti.module...

La première observation des données montre que nous avons un nombre considérable de colonnes (14470), mais la plupart d'entre elles semblent contenir des valeurs NaN, ce qui est logique puisque pandas étend les lignes comportant moins de colonnes afin qu'elles aient toutes le même nombre de colonnes, ajoutant ainsi des valeurs NaN aux champs manquants. Quelques utilisateurs ont gardé la même session pendant une très longue durée, alors que la plupart ont des sessions plus courtes ou de durée moyenne. Ainsi, les quelques sessions longues ont entraîné l'extension de toutes les autres à leur durée. C'est de là que proviennent toutes ces valeurs NaN.

In [36]:
# TODO: Inspecter les valeurs distinctes des navigateurs
print(features_train['browser'].unique())

['Firefox' 'Google Chrome' 'Microsoft Edge' 'Opera']


Quelques fonctions "utilitaires" qui pourraient vous être utiles (ou pas) :

In [38]:
import warnings
from IPython.display import display, Markdown

# décorateurs utilitaires pour supprimer les avertissements de la sortie et imprimer un cadre de données dans un tableau Markdown.
def ignore_warnings(f):
    def _f(*args, **kwargs):
        warnings.filterwarnings('ignore')
        v = f(*args, **kwargs)
        warnings.filterwarnings('default')
        return v
    return _f

# affiche un DataFrame Pandas sous forme de tableau Markdown dans un notebook Jupyter.
def markdown_table(headNtail=False, use_index=True, title=None, precision=2):
    def _get_value(val): return str(round(val, precision) if isinstance(val, float) else val)
    def _format_row(row): 
        row_str = ""
        if use_index: row_str += f"|{str(row.name)}"
        for value in row.values: row_str += f"| {_get_value(value)}"
        return row_str + "|"
    def _get_str(df):
        return "\n".join(df.apply(_format_row, axis=1))
    def _deco(f):
        def _f(*args, **kwargs):
            df = f(*args, **kwargs)
            _str = f"#### {title}\n" if title else ""
            header = ([str(df.index.name)] if use_index else []) + df.columns.astype(str).to_list() 
            _str += f"|{'|'.join(header)}|" + f"\n|{'--|'*len(header)}\n" if header else None
            if headNtail:
                _str += _get_str(df.head())
                _str += "\n|...|...|\n"
                _str += _get_str(df.tail())
            else:
                _str += _get_str(df)
            display(Markdown(_str))
        return _f
    return _deco

# fonction utilitaire permettant d'obtenir une grille graphique à partir d'un nombre arbitraire de lignes/colonnes ou de données.
def get_grid(n, n_row=None, n_col=None, titles=None, figsize=(10, 8), wspace=.5, hspace=.5, **kwargs):
    if n_row: n_col= n_col or math.floor(n/n_row)
    elif n_col: n_row= n_row or math.ceil(n/n_col)
    else:
        n_row = math.ceil(math.sqrt(n))
        n_col = math.floor(n/n_row)
    fig, axs = plt.subplots(n_row, n_col, figsize=figsize, **kwargs)
    plt.subplots_adjust(hspace=hspace, wspace=wspace)
    if titles is not None:
        for ax, title in zip(axs.flat, titles): ax.set_title(title)
    return fig, axs

Maintenant que nous avons un premier aperçu de nos données, approfondissons un peu et examinons quelques statistiques à leur sujet.

In [42]:
@markdown_table(title="Navigateur par utilisateur", headNtail=False)
def browsers_per_player(df):
    # Afficher ici pour chaque utilisateur (en ligne) le nombre de fois qu'il a utilisé chaque navigateur (en colonne)
    df_copy = df.copy()

    ohe = pd.get_dummies(df_copy['browser'], prefix='browser')
    # attach OHE columns
    df_copy = pd.concat([df_copy, ohe], axis=1)

    # list of OHE columns
    ohe_cols = [c for c in df_copy.columns if str(c).startswith('browser_')]

    # group by user and sum OHE columns (counts per user)
    users_browser = df_copy.groupby('util')[ohe_cols].sum().reset_index()

    # show result
    users_browser.head()
    return users_browser

# TODO: inspecter d'autre statistiques du jeu de données (la variable dépendante Y sera inspecté dans la prochaine cellule)
browsers_per_player(features_train)

#### Navigateur par utilisateur
|None|util|browser_Firefox|browser_Google Chrome|browser_Microsoft Edge|browser_Opera|
|--|--|--|--|--|--|
|0| aho| 0| 10| 0| 0|
|1| ajo| 13| 0| 0| 0|
|2| akx| 0| 14| 0| 0|
|3| ald| 0| 15| 0| 0|
|4| ats| 12| 0| 0| 0|
|5| avt| 15| 0| 0| 0|
|6| azn| 0| 4| 0| 0|
|7| azo| 0| 10| 0| 0|
|8| azx| 0| 10| 0| 0|
|9| bez| 0| 0| 4| 0|
|10| bfg| 16| 0| 0| 0|
|11| bgn| 27| 0| 0| 0|
|12| blf| 11| 0| 0| 0|
|13| bos| 9| 0| 0| 0|
|14| bou| 0| 11| 0| 0|
|15| btl| 0| 13| 0| 0|
|16| byo| 16| 0| 0| 0|
|17| cbo| 0| 13| 0| 0|
|18| cec| 10| 0| 0| 0|
|19| cil| 0| 8| 0| 0|
|20| cje| 0| 8| 0| 0|
|21| cjr| 0| 46| 0| 0|
|22| cka| 5| 0| 0| 0|
|23| ckl| 0| 0| 6| 0|
|24| cnh| 10| 0| 0| 0|
|25| cpn| 10| 0| 0| 0|
|26| cqp| 9| 0| 0| 0|
|27| crn| 0| 0| 4| 0|
|28| cum| 0| 0| 5| 0|
|29| cut| 18| 0| 0| 0|
|30| cuw| 0| 5| 0| 0|
|31| cvw| 12| 0| 0| 0|
|32| cxt| 0| 0| 15| 0|
|33| cxu| 0| 10| 0| 0|
|34| cyd| 0| 0| 18| 0|
|35| cyi| 0| 5| 0| 0|
|36| czi| 18| 0| 0| 0|
|37| dbi| 0| 0| 11| 0|
|38| deb| 10| 0| 0| 0|
|39| dfo| 11| 0| 0| 0|
|40| dit| 5| 0| 0| 0|
|41| ebm| 0| 9| 0| 0|
|42| eoe| 14| 0| 0| 0|
|43| eqo| 0| 5| 0| 0|
|44| eqw| 6| 0| 0| 0|
|45| esu| 14| 0| 0| 0|
|46| eus| 0| 0| 14| 0|
|47| fal| 12| 0| 0| 0|
|48| fcq| 0| 13| 0| 0|
|49| fek| 7| 0| 0| 0|
|50| fgc| 14| 0| 0| 0|
|51| fgn| 0| 0| 6| 0|
|52| fjb| 0| 0| 11| 0|
|53| fkf| 5| 0| 0| 0|
|54| flj| 0| 42| 0| 0|
|55| fmz| 0| 15| 0| 0|
|56| fou| 0| 21| 0| 0|
|57| ftq| 23| 0| 0| 0|
|58| fuy| 0| 15| 0| 0|
|59| fuz| 14| 0| 0| 0|
|60| fvc| 0| 29| 0| 0|
|61| fwf| 33| 0| 0| 0|
|62| fwh| 0| 13| 0| 0|
|63| fxg| 5| 0| 0| 0|
|64| fyf| 8| 0| 0| 0|
|65| fyg| 0| 0| 4| 0|
|66| gcr| 0| 0| 14| 0|
|67| gid| 0| 17| 0| 0|
|68| gor| 16| 0| 0| 0|
|69| gtj| 0| 19| 0| 0|
|70| guh| 0| 15| 0| 0|
|71| gvp| 0| 14| 0| 0|
|72| gzi| 6| 0| 0| 0|
|73| hdz| 8| 0| 0| 0|
|74| hew| 9| 0| 0| 0|
|75| hfy| 0| 7| 0| 0|
|76| hjs| 0| 0| 37| 0|
|77| hll| 0| 9| 0| 0|
|78| hrj| 0| 14| 0| 0|
|79| hvh| 0| 7| 0| 0|
|80| hxt| 0| 8| 0| 0|
|81| hzx| 0| 0| 20| 0|
|82| iba| 0| 8| 0| 0|
|83| ien| 0| 0| 9| 0|
|84| iib| 15| 0| 0| 0|
|85| iju| 0| 7| 0| 0|
|86| itt| 0| 4| 0| 0|
|87| iyf| 0| 8| 0| 0|
|88| jat| 0| 0| 5| 0|
|89| jew| 4| 0| 0| 0|
|90| jfm| 0| 18| 0| 0|
|91| jkc| 11| 0| 0| 0|
|92| jks| 0| 0| 8| 0|
|93| jkt| 0| 0| 4| 0|
|94| jmq| 0| 18| 0| 0|
|95| jpm| 21| 0| 0| 0|
|96| jqq| 0| 7| 0| 0|
|97| jrc| 9| 0| 0| 0|
|98| jxu| 10| 0| 0| 0|
|99| kbf| 6| 0| 0| 0|
|100| khv| 13| 0| 0| 0|
|101| kik| 0| 15| 0| 0|
|102| kks| 27| 0| 0| 0|
|103| kwi| 31| 0| 0| 0|
|104| kzt| 13| 0| 0| 0|
|105| lfq| 5| 0| 0| 0|
|106| lgd| 15| 0| 0| 0|
|107| lgz| 25| 0| 0| 0|
|108| lhd| 7| 0| 0| 0|
|109| lhr| 12| 0| 0| 0|
|110| lio| 17| 0| 0| 0|
|111| lnq| 5| 0| 0| 0|
|112| lra| 0| 0| 4| 0|
|113| meh| 16| 0| 0| 0|
|114| mgs| 12| 0| 0| 0|
|115| mmm| 15| 0| 0| 0|
|116| mmo| 0| 5| 0| 0|
|117| mqo| 0| 0| 9| 0|
|118| muz| 0| 28| 0| 0|
|119| mvu| 0| 5| 0| 0|
|120| myl| 0| 11| 0| 0|
|121| ncb| 0| 20| 0| 0|
|122| nfh| 0| 6| 0| 0|
|123| ngs| 0| 16| 0| 0|
|124| nkb| 0| 7| 0| 0|
|125| nlm| 15| 0| 0| 0|
|126| npo| 6| 0| 0| 0|
|127| nqg| 0| 0| 34| 0|
|128| nrb| 15| 0| 0| 0|
|129| nsf| 0| 6| 0| 0|
|130| nuh| 11| 0| 0| 0|
|131| nwt| 0| 20| 0| 0|
|132| nyk| 15| 0| 0| 0|
|133| nzq| 26| 0| 0| 0|
|134| obm| 8| 0| 0| 0|
|135| oec| 19| 0| 0| 0|
|136| ofq| 11| 0| 0| 0|
|137| oni| 7| 0| 0| 0|
|138| ont| 0| 0| 5| 0|
|139| oph| 0| 0| 7| 0|
|140| opp| 24| 0| 0| 0|
|141| oqa| 0| 15| 0| 0|
|142| pcv| 0| 15| 0| 0|
|143| pgi| 0| 25| 0| 0|
|144| pgm| 20| 0| 0| 0|
|145| pgu| 18| 0| 0| 0|
|146| pou| 7| 0| 0| 0|
|147| pqd| 9| 0| 0| 0|
|148| prb| 16| 0| 0| 0|
|149| pre| 0| 16| 0| 0|
|150| psg| 0| 8| 0| 0|
|151| ptb| 16| 0| 0| 0|
|152| pxi| 7| 0| 0| 0|
|153| pyb| 10| 0| 0| 0|
|154| qbf| 0| 12| 0| 0|
|155| qdv| 0| 5| 0| 0|
|156| qfy| 0| 0| 12| 0|
|157| qkd| 0| 0| 15| 0|
|158| qof| 0| 16| 0| 0|
|159| qqe| 0| 15| 0| 0|
|160| qsw| 0| 20| 0| 0|
|161| rff| 4| 0| 0| 0|
|162| rgo| 0| 0| 17| 0|
|163| rmf| 0| 0| 6| 0|
|164| rvh| 0| 14| 0| 0|
|165| sbz| 0| 5| 0| 0|
|166| sdx| 0| 11| 0| 0|
|167| sfi| 0| 16| 0| 0|
|168| sgd| 7| 0| 0| 0|
|169| shh| 34| 0| 0| 0|
|170| sii| 19| 0| 0| 0|
|171| skm| 0| 75| 0| 0|
|172| slq| 0| 71| 0| 0|
|173| snn| 0| 0| 8| 0|
|174| snv| 21| 0| 0| 0|
|175| sov| 0| 6| 0| 0|
|176| sps| 0| 0| 20| 0|
|177| sqb| 0| 9| 0| 0|
|178| swe| 0| 0| 0| 23|
|179| syt| 16| 0| 0| 0|
|180| szr| 15| 0| 0| 0|
|181| tdk| 14| 0| 0| 0|
|182| tfi| 13| 0| 0| 0|
|183| tft| 18| 0| 0| 0|
|184| tjc| 0| 9| 0| 0|
|185| tkz| 0| 8| 0| 0|
|186| tmn| 6| 0| 0| 0|
|187| tsn| 21| 0| 0| 0|
|188| tvt| 9| 0| 0| 0|
|189| txk| 16| 0| 0| 0|
|190| tyw| 0| 0| 13| 0|
|191| tza| 15| 0| 0| 0|
|192| tzn| 0| 12| 0| 0|
|193| ubr| 0| 6| 0| 0|
|194| udj| 0| 0| 10| 0|
|195| uey| 0| 18| 0| 0|
|196| ufc| 0| 5| 0| 0|
|197| uhx| 5| 0| 0| 0|
|198| ulu| 0| 0| 15| 0|
|199| uof| 0| 8| 0| 0|
|200| urs| 0| 5| 0| 0|
|201| usg| 10| 0| 0| 0|
|202| usy| 6| 0| 0| 0|
|203| uwp| 12| 0| 0| 0|
|204| uwr| 15| 0| 0| 0|
|205| vek| 16| 0| 0| 0|
|206| vfo| 0| 35| 0| 0|
|207| vjw| 0| 12| 0| 0|
|208| vlm| 0| 26| 0| 0|
|209| vmy| 0| 7| 0| 0|
|210| vog| 5| 0| 0| 0|
|211| vos| 5| 0| 0| 0|
|212| vsb| 11| 0| 0| 0|
|213| vyu| 7| 0| 0| 0|
|214| vzc| 0| 12| 0| 0|
|215| wcf| 9| 0| 0| 0|
|216| wem| 0| 15| 0| 0|
|217| wfs| 0| 18| 0| 0|
|218| wht| 0| 20| 0| 0|
|219| wjn| 0| 0| 21| 0|
|220| wuy| 0| 15| 0| 0|
|221| wvy| 12| 0| 0| 0|
|222| wwk| 0| 12| 0| 0|
|223| wxf| 8| 0| 0| 0|
|224| xby| 14| 0| 0| 0|
|225| xgb| 7| 0| 0| 0|
|226| xgv| 22| 0| 0| 0|
|227| xkr| 0| 14| 0| 0|
|228| xvr| 14| 0| 0| 0|
|229| xwz| 0| 6| 0| 0|
|230| xxb| 5| 0| 0| 0|
|231| ycb| 0| 11| 0| 0|
|232| ykq| 0| 0| 7| 0|
|233| yrb| 0| 0| 10| 0|
|234| yuz| 15| 0| 0| 0|
|235| yvq| 0| 13| 0| 0|
|236| yxw| 5| 0| 0| 0|
|237| zbn| 15| 0| 0| 0|
|238| zct| 0| 14| 0| 0|
|239| zht| 14| 0| 0| 0|
|240| zll| 0| 0| 4| 0|
|241| zok| 0| 0| 11| 0|
|242| zqs| 0| 0| 16| 0|
|243| zro| 13| 0| 0| 0|
|244| zrx| 0| 0| 12| 0|
|245| zus| 13| 0| 0| 0|
|246| zyk| 0| 21| 0| 0|

// Votre analyse ici

In [ ]:
@markdown_table(headNtail=True, title="Stats: Y distribution")
def get_Y_stats(df):
    # Inspecter la distribution de la variable dépendante ici (Y)

// Votre analyse ici

## Construction de caractéristiques

### Traitements préliminaires

En fonction du modèle choisi, différentes étapes de pré-traitement peuvent être nécessaire. Voir https://scikit-learn.org/stable/supervised_learning.html pour des exemples de modèles dans le cas supervisé (notre cas). Quelques exemples : 
- SVMs
- Régression logistiques
- Arbres de décision
- Modèle ensembliste (Random Forest, XGBoost, etc.)
- Réseaux de neurones 

Il faut commencer par se demander s'il y a des valeurs aberrantes (*outliers*) et, le cas échéant, appliquer le traitement approprié (pour le moment vous pouvez à priori les supprimer).

In [ ]:
# TODO: inspecter s'il y a des valeurs aberrantes, et les enlever le cas échéant

### Variable de classe
Notre variable dépendante est une chaîne de caractères (str). Nous pouvons la convertir en codes catégoriels (numériques) à l'aide de la fonction pd.Categorical.

In [ ]:
features_train["util"] = pd.Categorical(features_train["util"])

pd.Categorical ne modifie pas directement l'ID utilisateur en un nombre, mais lui ajoute un attribut cat.codes. Nous pouvons créer une petite fonction pour convertir la variable de classe d'une chaîne de caractères en son ID de catégorie :

In [ ]:
def to_categories(df, col="util"):
    df[[col]] = df[[col]].apply(lambda x: x.cat.codes)

### Obtention des caractéristiques...

Nous allons maintenant créer des caractéristiques à partir de l'ensemble de données. Pour cela, il va falloir comprendre leur format, les parser et voir comment les agréger. On peut commencer par inspecter toutes les actions différentes, indépendemment des informations sur l'écran ou la fiche travaillée.

In [ ]:
def filter_action(value: str):
    for delim in [ "(", "<", "$", "1"]:
        if delim in value and (low_ind := value.index(delim)):
            value = value[:low_ind]
    return value

uniques = features_train.iloc[:,2:].stack().unique()
filtered_uniques = list(set([filter_action(un) for un in uniques if not un.startswith("t")]))
len(filtered_uniques), filtered_uniques

In [ ]:
#TODO Définir des features sur les données brutes et les placer dans un nouveau dataframe

Ensuite, on peut employer des fonctions plus complexes pour extraire l'écran le plus utilisé, la configuration d'écran la plus utilisé et la chaîne (catégorie) de la fiche la plus utilisé. Pour vous aider, voici les expressions régulières (*regex*) qui permettent d'extraire de telles informations. N'hésitez pas à extraire d'autres informations si cela vous parait pertinent.

In [ ]:
import re
from collections import Counter

pattern_ecran = re.compile(r"\((.*?)\)")
pattern_conf_ecran = re.compile(r"<(.*?)>")
pattern_chaine = re.compile(r"\$(.*?)\$")


### Traitement des chaînes de caractère
La colonne navigateur ne peut prendre que quatre valeurs ; pour la convertir en nombre, on a deux choix : 
- la convertir en variable catégorielle comme nous l'avons fait avec notre variable dépendante
- la convertir en One-Hot Encoding (OHE), un pour chaque navigateur. Pandas propose également une fonction pour cela : get_dummies

En fonction du modèle, l'un ou l'autre peut être plus pertinent (par exemple, pour un arbre de décision ?).

In [ ]:
#TODO : traiter les chaînes de caractères

## Statistiques simples sur les variables

### Inspection des caractéristiques

Commencez par inspecter les caractéristiques que vous avez construites, et afficher des graphiques en les interprétant. Vous pouvez utiliser la librairie Matplotlib à cette fin.

In [ ]:
fig, (ax1, ax2, ax3) = plt.subplots(ncols=3, sharey=True, figsize=(20,6))
# TODO: afficher des graphiques de trois features différentes sur les trois axes donnés

// Votre analyse ici

### Inspection de la corrélation d'une caractéristique pour une classe particulière
Comme il peut être difficile d'inspecter les données lorsque nous avons un tel nombre de classes possibles, nous pouvons commencer par tracer des graphiques de corrélation pour une ou quelques cibles particulières, choisies arbitrairement.


In [ ]:
def plot_categories(categories, type_="box", feature="actions_means"):
    fig, axs = get_grid(len(categories), n_row=1, figsize=(16, 4))
    # TODO : afficher les graphiques des catégories passées en paramètre

In [ ]:
categories = [0, 15, 152, 199]
plot_categories(categories)

Un boxplot est très utile pour estimer la distribution d'une caractéristique au sein d'une catégorie (cf. cours), mais d'autres visualisations peuvent servir comme les violin plots, cf. : https://www.data-to-viz.com



// à compléter

## Développement d'une solution de classification automatique

Une fois que vous avez exploré les données et mieux compris comment résoudre la tâche qui vous est confiée, c'est le moment de commencer à développer votre solution de classification automatique supervisée.

La première opération consiste à séparer votre jeu de données en deux sous-ensembles : un jeu d'entraînement (**training set**) et un jeu de validation (**validation set**). Le jeu de validation vous aidera dans la sélection du meilleur modèle. Suivant les algorithmes employés, et leur coût computationnel, vous pourriez aussi avoir recours à la validation croisée à K-folds (*K-fold cross validation*).

Il s'agit ensuite de tester plusieurs algorithmes classiques et de faire varier ses hyper-paramètres, tel que :

- Régression logistique
- Machines à vecteurs supports (SVM) : différents types de noyau (linéaire, polynomial à différents degrés)
- Arbres de décisions simples
- Ensembles d'arbres (Random Forest, XGBoost) : nombre d'arbres
- Réseaux de neurones artificiels (simple MLP) : nombre de couches, nombre de neurones par couche

Vous pouvez également tester les méthodes de régularisation afin d'obtenir des modèles parcimonieux.

Pour cela, vous privilégierez l'usage de la librairie *scikit-learn*.

## Evaluation de votre solution

Les algorithmes déployés dans la section précédente permettent de réaliser des prédictions sur des données qui n'ont pas été vus durant l'entraînement. Dans le cadre de cette compétition, vous utiliserez pour ça un ensemble de validation.

Il s'agit à présent de définir convenablement et d'utiliser les métriques les plus appropriées à votre tâche. On vous conseille à minima d'employer la **réussite simple** en classification (*accuracy*) et la **F-Mesure** (ou F1-score) mais vous êtes aussi encouragés à utiliser d'autres manière d'évaluer l'efficacité d'un modèle.

// à compléter

## Discussion au sujet de vos résultats

Il faut discuter de vos résultats, si possible en prenant en compte tous les critères que vous jugerez utiles (cf. mesures discutées dans la partie précédente).

## Evaluation sur l'ensemble de données de test

Il s'agit enfin d'utiliser votre modèle sur l'ensemble de données de test fourni.

In [ ]:
#TODO: preprocess test dataset, compute features values and post-process it (handle string, etc.)
test_processed_df = # TODO

In [ ]:
preds_test =  votre_modele.predict(test_processed_df)
preds_test

## Soumission à la compétition Kaggle

Il faut bien mettre en forme le fichier que vous pourrez utiliser pour soumettre sur le site de la compétitiokn.

In [ ]:
preds_test = pd.Categorical.from_codes(preds_test, features_train["util"].cat.categories)
df_subm = pd.DataFrame(preds_test)
df_subm = df_subm.rename_axis('RowId')
df_subm.rename(columns={0: 'prediction'}, inplace=True)
df_subm.index = df_subm.index + 1
df_subm

In [ ]:
df_subm.to_csv('submission.csv')